### PR TITLE
SQLmap supports clean URLs

### PR DESCRIPTION
**Pull request by kazet**

### CODE REVIEW

File: karton_sqlmap/karton_sqlmap.py [hunk 1]
```diff
@@ -102,37 +112,151 @@ def run(self, current_task: Task) -> None:
...
+                                    url, sqlmap_options, find_in_output, timeout_seconds=SQLI_ADDITIONAL_DATA_TIMEOUT
+                                ),
+                            )
                         except Exception:  # Whatever happens, we prefer to report SQLi without additional data than no SQLi
                             self.log.exception(f"Unable to obtain {information_name} via blind SQL injection")
+                    return found_sql_injection
+        return None
+
+    @staticmethod
+    def _expand_query_parameters_for_scanning(url: str) -> List[str]:
```
Comment (bulkowy) [inline karton_sqlmap/karton_sqlmap.py:145]: I would like to see example in some sort of comment, code is not very self-explanatory

File: karton_sqlmap/karton_sqlmap.py [hunk 2]
```diff
@@ -102,37 +112,151 @@ def run(self, current_task: Task) -> None:
...
+        results = []
+        for key in query:
+            new_query = copy.copy(query)
+            token = "__sqlmap_injection_point__"
+            for item in [new_query[key] + token, token]:
+                new_query[key] = item
+
+                # We replace token with * after building the URL, so that the asterisk is passwd to sqlmap unescaped
+                results.append(
+                    urllib.parse.urlunparse(url_parsed._replace(query=urllib.parse.urlencode(new_query))).replace(
```
Comment (bulkowy) [inline karton_sqlmap/karton_sqlmap.py:161]: don't stack such functions, execute them every line, so it's is clearer to read

File: karton_sqlmap/karton_sqlmap.py [hunk 3]
```diff
@@ -102,37 +112,151 @@ def run(self, current_task: Task) -> None:
...
+                        token, "*"
+                    )
+                )
+        return results
+
+    @staticmethod
+    def _expand_path_segments_for_scanning(url: str) -> List[str]:
```
Comment (bulkowy) [inline karton_sqlmap/karton_sqlmap.py:168]: same as above, needs comments

**Review by bulkowy — APPROVED** [review]

File: karton_sqlmap/karton_sqlmap.py [hunk 4]
```diff
@@ -102,37 +112,188 @@ def run(self, current_task: Task) -> None:
...
+        num_slashes = len([c for c in url_parsed.path[1:] if c == "/"])
+
+        if num_commas > num_slashes:
+            separator = ","
+        else:
+            separator = "/"
+
+        extension_re = r"\.[a-z]{3,4}$"
+        if m := re.search(extension_re, url_parsed.path):
+            extension = url_parsed.path[m.start() : m.end()]
```
Comment (bulkowy) [inline karton_sqlmap/karton_sqlmap.py:202]: you dont need to retrieve it, as regex did it before

File: karton_sqlmap/karton_sqlmap.py [hunk 5]
```diff
@@ -102,37 +112,188 @@ def run(self, current_task: Task) -> None:
...
+            path_segments = url_parsed.path[1 : -len(extension)].split(separator)
```
Comment (bulkowy) [inline karton_sqlmap/karton_sqlmap.py:203]: and then you can just do it by getting len of matched element

File: karton_sqlmap/karton_sqlmap.py [hunk 6]
```diff
@@ -102,37 +112,221 @@ def run(self, current_task: Task) -> None:
...
+                        urls_with_injection_points.add(url_with_injection_point)
+                        expanded_urls_with_example_values.append(
+                            (
+                                url_with_injection_point,
+                                example_value,
+                            )
+                        )
+
+        random.shuffle(expanded_urls_with_example_values)
+        expanded_urls_with_example_values = expanded_urls_with_example_values[
```
Comment (bulkowy) [inline karton_sqlmap/karton_sqlmap.py:295]: I would like for loop to end if len(expanded_urls) > config_value, no need for it to loop through stuff taht would be truncated anyway

Comment (kazet) [reply karton_sqlmap/karton_sqlmap.py:295]: but we perform random.shuffle() on that to get a random sample of the urls

without that, we wouldn't consider the urls that appear at the bottom of the page

and this loop is fast, it doesn't perform any server interaction

**Review by bulkowy — APPROVED** [review]

**Pull request merged**